### PR TITLE
Layernorm

### PR DESCRIPTION
**Pull request by OneAdder**
# Layer Normalization

Layer Normalization applies normalization over a sequence of inputs by the following formula:

$\frac{x - \bar{x}}{\sqrt{var(x) + \epsilon}} \cdot \gamma + \beta$

Where:
* $\bar{x}$ is arithmetic mean over the last dimension
* $var(x)$ is variance over the last dimension
* $\epsilon$ is a scaling constant (small)
* $\gamma$ and $\beta$ are trainable parameters

### Reference

Lei Ba, J., Kiros, J. R., & Hinton, G. E. (2016). Layer normalization. ArXiv e-prints, [arXiv-1607](https://arxiv.org/pdf/1607.06450).

### Status

Ready

### CODE REVIEW

File: src/nf/nf_layernorm_submodule.f90 [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    ! normalize mu by variance by first axis
+    do concurrent(i = 1: self % model_dimension)
+      normalized(:, i) = self % mu(:, i) / self % sigma
+    end do
+
+    ! forward through trainable params gamma and beta
+    do concurrent(i = 1: self % sequence_length)
+      self % output(i, :) = normalized(i, :) * self % gamma + self % beta
+    end do
```
Comment (jvdp1) [inline src/nf/nf_layernorm_submodule.f90:35]: It seems that these two loops could be combined, such that the variable `normalized` would not be needed.

File: test/test_layernorm.f90 [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  call layernorm % init([3, 4])
+
+  call test_layernorm_forward(layernorm, sample_input, ok)
+  call test_layernorm_backward(layernorm, sample_input, sample_gradient, ok)
+
+  if (ok) then
+    print '(a)', 'test_layernorm_layer: All tests passed.'
+  else
+    write(stderr, '(a)') 'test_layernorm_layer: One or more tests failed.'
+    stop 1
```
Comment (jvdp1) [inline test/test_layernorm.f90:21]:
```suggestion
    error stop 1
```

File: src/nf/nf_layernorm_submodule.f90 (hunk 1 above)
Comment (OneAdder) [reply src/nf/nf_layernorm_submodule.f90:35]: Yes, I'll fix that

Comment (OneAdder) [reply src/nf/nf_layernorm_submodule.f90:35]: Can you explain how would you do that? The only thing I can think of is making a nested loop like this:
```fortran
do concurrent(i = 1: self % sequence_length, j = 1: self % model_dimension)
```
But that'll obviously make it $n \cdot m$ complexity

Comment (jvdp1) [reply src/nf/nf_layernorm_submodule.f90:35]: I was indeed thinking to such a nested loop, as it is currently twice the case implicitely. but I might be wrong, and this could be checked at a future stage.

Comment (OneAdder) [reply src/nf/nf_layernorm_submodule.f90:35]: In this case I'd definitely use some extra memory to reduce processor complexity to $n + m$

**Comment by OneAdder** [comment]
@milancurcic @jvdp1 Ready for review

File: src/nf/nf_layernorm_submodule.f90 [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    res % eps = 1e-5
+  end function layernorm_layer_cons
+
+  pure module subroutine forward(self, input)
+    class(layernorm_layer), intent(in out) :: self
+    real, intent(in) :: input(:, :)
+    real, allocatable :: normalized(:, :)
+    integer :: i
+
+    allocate(normalized(self % sequence_length, self % model_dimension))
```
Comment (jvdp1) [inline src/nf/nf_layernorm_submodule.f90:17]:
```suggestion
    allocate(normalized, mold=self % mu)
```

**Review by jvdp1 — APPROVED** [review]
LGTM @OneAdder . Nice work!

File: src/nf/nf_layer_constructors.f90 [hunk 4]
```diff
@@ -222,15 +223,23 @@ module function linear2d(out_features) result(res)
...
+      !! Rank-2 (sequence_length, out_features) self attention constructor.
+      !! sequence_length and model_dimension are determined at layer initialization, based on the
+      !! output shape of the previous layer.
+      integer, intent(in) :: num_heads
+        !! Number of attention heads
+      type(layer) :: res
+        !! Resulting layer instance
+    end function self_attention
+
+    module function layer_normalization() result(res)
```
Comment (milancurcic) [inline src/nf/nf_layer_constructors.f90:236]: How about `layernorm` for the constructor? Shorter, and `norm` is quite ubiquitous to mean "normalization". PyTorch uses `LayerNorm` and Keras uses `LayerNormalization`.

File: src/nf/nf_layernorm_submodule.f90 [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    res % eps = 1e-5
+  end function layernorm_layer_cons
+
+  pure module subroutine forward(self, input)
+    class(layernorm_layer), intent(in out) :: self
+    real, intent(in) :: input(:, :)
+    real, allocatable :: normalized(:, :)
+    integer :: i
+
+    allocate(normalized, mold=self % mu)
```
Comment (milancurcic) [inline src/nf/nf_layernorm_submodule.f90:17]: Since the shape of `mu` is determined at `init`, so is the shape of `normalized`.  It's inefficient to allocate+deallocate an array in every `forward` call. Let's allocate `normalized` in `init` as well.

File: src/nf/nf_layernorm_submodule.f90 [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+  pure module subroutine backward(self, input, gradient)
+    class(layernorm_layer), intent(in out) :: self
+    real, intent(in) :: input(:, :)
+    real, intent(in) :: gradient(:, :)
+    real, allocatable :: one_over_sigma(:, :)
+    real, allocatable :: gradient_by_gamma_over_sigma(:, :)
+
+    allocate(one_over_sigma(self % sequence_length, self % model_dimension))
+    allocate(gradient_by_gamma_over_sigma(self % sequence_length, self % model_dimension))
```
Comment (milancurcic) [inline src/nf/nf_layernorm_submodule.f90:48]: Same for these, should be allocated at `init` time

File: test/test_layernorm.f90 [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  use iso_fortran_env, only: stderr => error_unit
+  use nf_layernorm_layer, only: layernorm_layer
+  use nf, only: sgd
+  implicit none
+
+  logical :: ok = .true.
+  type(layernorm_layer) :: layernorm
+  real :: sample_input(3, 4) = reshape([0.0, 10.1, 0.2, 10.3, 0.4, 10.5, 0.6, 10.7, 10.8, 0.9, 0.11, 0.12], [3, 4])
+  real :: sample_gradient(3, 4) = reshape([0.1, 3., 2., 0.1, 3., 3., 0.1, 2., 0.1, 3., 0.1, 3.], [3, 4])
+
```
Comment (milancurcic) [inline test/test_layernorm.f90:11]: Test the constructor too.

**Comment by milancurcic** [comment]
Great, thank you! I left several comments.

File: test/test_layernorm.f90 [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+  layernorm = layernorm_layer()
+  call layernorm % init([3, 4])
+
+  call test_layernorm_forward(layernorm, sample_input, ok)
+  call test_layernorm_backward(layernorm, sample_input, sample_gradient, ok)
+  call test_layernorm_gradients(sample_input, sample_gradient, ok)
+
```
[the remark this answers is not in the document]
Comment (OneAdder) [reply test/test_layernorm.f90:21]: @milancurcic I have made such test here. The only issue here is that margin for error should be rather high (`0.1`) and a lot of training epochs. The issue here is that to test convergence we use LayerNormalization as an activation function which works but is inefficient

File: test/test_layernorm.f90 [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    integer :: epochs = 10000
+
+    net = network([&
+        input(2, 3),&
+        linear2d(3),&
+        layernorm(),&
+        flatten()&
+    ])
+
+    ! Kaiming weights to achieve semblance of convergance
```
Comment (OneAdder) [inline test/test_layernorm.f90:172]: If we want LayerNorm used as activation function to converge, we need to at least use correct weights initialization. Otherwise, the net will not be able to consistently converge and we'll get https://github.com/modern-fortran/neural-fortran/issues/145 once more

File: README.md [hunk 10]
```diff
@@ -34,8 +34,9 @@ Read the paper [here](https://arxiv.org/abs/1902.06714).
 | Dropout | `dropout` | `dense`, `flatten`, `input1d` | 1 | ✅ | ✅ |
 | Convolutional (2-d) | `conv2d` | `input3d`, `conv2d`, `maxpool2d`, `reshape` | 3 | ✅ | ✅(*) |
 | Max-pooling (2-d) | `maxpool2d` | `input3d`, `conv2d`, `maxpool2d`, `reshape` | 3 | ✅ | ✅ |
-| Linear (2-d) | `linear2d` | `input2d`, `linear2d`, `self_attention` | 2 | ✅ | ✅ |
-| Self-attention | `self_attention` | `input2d`, `linear2d`, `self_attention` | 2 | ✅ | ✅ |
+| Linear (2-d) | `linear2d` | `input2d`, `layernorm`, `linear2d`, `self_attention` | 2 | ✅ | ✅ |
+| Self-attention | `self_attention` | `input2d`, `layernorm`, `linear2d`, `self_attention` | 2 | ✅ | ✅ |
+| Layer Normalization | `layernorm` | `linear2d`, `self_attention` | 2 | ✅ | ✅ |
```
Comment (milancurcic) [inline README.md:39]: @OneAdder can you please check that I did this correctly?

**Review by milancurcic — APPROVED** [review]
Thank you!!

File: README.md (hunk 10 above)
Comment (OneAdder) [reply README.md:39]: Yes, looks good!
